### ISSUE TITLE
Medtronic pumps (maybe pods too?) continue to “dose insulin” after reservoir if empty.

### ISSUE THREAD

**Issue by tqb43** [Artificial-Pancreas/iAPS#241]
A user on Facebook noted that she fell asleep and her Medtronic reservoir hit 0. She slept for about 6 hours with the empty reservoir but during that time iAPS continually calculated and “delivered” basals and SMBs. 

Obviously, she wasn’t getting that insulin but she woke up with a massive amount of false IOB. 

I suggest once a reservoir hits 0 to put the pump into suspend mode.

not sure if pods behave similarly.

**Comment by LiroyvH** [comment]
I think you actually *do* get that insulin on Medtronic pumps though. I've seen similar behavior with my Medtronic 754 and 722's: both can say 0 units remaining, and iAPS shows the same, but it continues delivering insulin. You can hear and see the motor moving and when I detach myself from the cable: insulin comes dripping out. At some point, it truly comes to the end and then it throws a Motor Error/Fault. Typically I think once it hits 0U I have at least ~20-25U remaining. Once I get a Motor Error I think iAPS jumped to Red Loop as well as I got "iAPS has not been active for xx minutes blabla". 

So with that in mind, is she absolutely 100% positive it was "false IOB" and not actually delivered insulin...? The Medtronic saying 0U remaining is more of a safe-side estimation than that it seems to be true. I'm quite liking this margin tbh.

**Comment by Chester-sk8** [comment]
When I used 722 pump without iaps it were about 20 units in reservoir after 0 in pump’s info

**Comment by LiroyvH** [comment]
Hi, 

I tried again, but cannot reproduce this problem. Here are all the details:

**Setup**:
- App : iAPS 2.2.5-dev 
- Pump: Medtronic 754 v2.6A1.1
  -- Insulin: Lyumjev
  -- Battery: Lithium
  -- Data source: Event History
  -- Use MySentry: No
- Link: EmaLink-Medtronic, fw 2.12/3.7

**Expected**: 
pump stops delivery once truly empty and throws error, iAPS must error out subsequently. 

**Actual result**:
The Medtronic says 0U remaining, but continues delivering. Once it's really empty, it throws a motor error, including audible alert, and suspends insulin delivery. 

The result of that is:
1.) iAPS properly corrected the delivered bolus from 5U to 3.3U actually delivered (before failure state)
2.) iAPS upon subsequent loops throws error "Pump suspended" when it tries to SMB or manual bolus. 
3.) After 20 minutes, it starts throwing "last loop X minutes ago" notifications. 

_NOTE:_ This process was done, albeit by accident, on prior occasion with a Medtronic 722 with an EmaLink and OrangeLink connected, with the exact same configuration and result.


**Conclusion**: 
Cannot reproduce, pump and iAPS behave exactly as expected with proper errors, sounds, notifications, IOB and bolus history. Maybe the setup is different, then more information is required and advisable to try and repeat the failure state.

**Comment by csidrak** [comment]
@LiroyvH Thank you for testing this issue out. I'm not sure which of my iaps settings, if any, may have contributed to the experience I had with iaps continuing to deliver SMB while my reservoir was empty, but suffice it say that I'm keeping a close watch on my pump reservoir volume before I go to bed now!

**Comment by LiroyvH** [comment]
@csidrak Yeah I can imagine so. The problem is pretty much that the pump reports back to iAPS that it's still delivering insulin; and in my experience and testcase, and by the looks of it also in @Chester-sk8's case, it actually *does* deliver it until it errors out. I seem to recall that a few years ago, Katie DiSimone (RIP) also wrote about this phenomenon and how the Medtronic's always estimate the 0U remaining wrong - but the documentation of Loop has been overhauled so many times and relocated multiple times, I can't find it anymore.

If you're able to test it sometime, for example by - when it says 0U - disconnecting from your body and then trying to bolus like 5U a couple of times and see what it does (insulin dripping out?) (you can delete it from treatments later) and whether or not it errors out; that'd be helpful. Stay safe though!

**Comment by Chester-sk8** [comment]
In addition can say that when you fill reservoir with 300U, medtronic’s pump and iaps show only 280u. This is exactly those “lost” 20 units which you can use when iaps shows 0

**Comment by csidrak** [comment]
@LiroyvH, I switched to Dash pods 3 days ago so I don’t think I have access to the pump event history for Oct 9 (which was the day I had the issue with iAPS delivering non-existing insulin to my medtronic pump via SMB). I may go back to medtronic at some point (I still have probably 5-6 months of supplies), but needed a break from some of the other issues I had with it, mainly the frequent yellow and red loops which seems to be expected if I continue looping with medtronic and iAPS.

**Comment by csidrak** [comment]
@Chester-sk8 your English is great. I would have never thought there could be a 20 unit difference in these pumps. I wonder if it is specific to certain medtronic pump models? I’ve only ever used 522 and 722 medtronic pump models.

**Comment by Chester-sk8** [comment]
> @Chester-sk8 your English is great! I would have never thought there could be a 20 unit discrepancy in medtronic pumps. I wonder if it is specific to certain medtronic pump models? I’ve only ever used 522 and 722 medtronic pump models.

I’ve only used 722 pump and as I see this is usual situation when I fill a new reservoir with 300 units (I have insulin in flexpen with 300 units) and pump shows in statistic only about 280-283 units. For filling reservoir I use about 7-9 units. Yesterday I refilled my already used reservoir which contained some numver of units by new insulin (bad way but I wanted to test) and iaps shown 302 units after start but in a minute has changed it to 282:)

**Comment by LiroyvH** [comment]
@csidrak Alright :) Just let us know if you do go back to Medtronic again and end up testing it. And you're welcome!

Also, not sure why when using Medtronic you should be expecting frequent yellow/red loops. I'm using iAPS with a Medtronic as well and it's quite rare to get yellow and even rarer to get red (other than when suspending for shower and such, but that doesn't count. :)) Maybe if you do go back to Medtronic we can look at troubleshooting that as well. 

Enjoy the Dash! :)

**Comment by csidrak** [comment]
I’ve been on medtronic for so long for my pumping years that it actually
feels weird not to have tubing and I didn’t mind carrying around the orange
link. But I just was constantly troubleshooting the yellow and red loops
almost daily. I had similar issues when I used medtronic and orange link
for DIY Loop app back in May of this year, so I figured the communications
between the orange link to the pump was just not great.

This is everything I tried:

-  I used fresh lithium batteries in my pump and orange link periodically

- Downloaded the most up to date orangelink firmware

- Rarely kept my phone on low power mode

- almost always had my orange link, pump and phone on my person (with me)

-  closed and reopened the iAPS app, restarted phone sometimes

I tried all of the above with both my 522 and 722 pumps and still had
issues.

Is there anything else you recommend I try?

(I can switch back and forth between Dash and medtronic as long as I
remember to change my basal rates.)

-  Christina
